stdlib/Printf/src/Printf.jl

```diff
@@ -379,16 +379,30 @@ tofloat(x) = Float64(x)
 tofloat(x::Base.IEEEFloat) = x
 tofloat(x::BigFloat) = x
 
+_snprintf(ptr, siz, str, arg) =
+    @ccall "libmpfr".mpfr_snprintf(ptr::Ptr{UInt8}, siz::Csize_t, str::Ptr{UInt8};
+                                   arg::Ref{BigFloat})::Cint
+
 @inline function fmt(buf, pos, arg, spec::Spec{T}) where {T <: Floats}
     leftalign, plus, space, zero, hash, width, prec =
         spec.leftalign, spec.plus, spec.space, spec.zero, spec.hash, spec.width, spec.precision
     x = tofloat(arg)
-    if x isa BigFloat && isfinite(x)
-        ptr = pointer(buf, pos)
-        newpos = @ccall "libmpfr".mpfr_snprintf(ptr::Ptr{UInt8}, (length(buf) - pos + 1)::Csize_t, string(spec; modifier="R")::Ptr{UInt8}; arg::Ref{BigFloat})::Cint
-        newpos > 0 || error("invalid printf formatting for BigFloat")
-        return pos + newpos
-    elseif x isa BigFloat
+    if x isa BigFloat
+        if isfinite(x)
+            GC.@preserve buf begin
+                siz = length(buf) - pos + 1
+                str = string(spec; modifier="R")
+                len = _snprintf(pointer(buf, pos), siz, str, arg)
+                if len > siz
+                    len > 4096 &&
```

Comment:
This cutoff seems pretty small to me --- 4k is not an unreasonable amount of data.

Reply:
That's for a single BigFloat, which would mean a precision of over 13600 bits to hit that limit (I hadn't heard of anybody using more than 1024 bits for BigFloat, so I thought that 4096 would be large enough (and definitely better than the current behavior of just giving a garbage truncated result with lots of  \0 bytes)
I can change that to use something like the `min(__BIGFLOAT_LIMIT__, ceil(Int, precision(arg)*log(2)/log(10)) + 32)`, so that it would always be able to output (in a round-trippable fashion) any BigFloat using `%e` or `%g`, as well as handle any size up to __BIGFLOAT_LIMIT__ for `%f` whatever you think that should be set to, because of the 64-bit binary exponent having to be output as 0's or extra digits).